// react-native-fxview/FXView.tsx
import React, { Component } from "react";
import { View, ViewProps } from "react-native";
import { FXComponentItem } from "./types";
import FXManager from "./FXViewManager";
import logger from "./logger/FXLogger";

interface FXViewProps extends ViewProps {
  fxViewId?: string; // 动态标识属性，可选
}

interface FXViewState {
  components: FXComponentItem[];
}

export default class FXView extends Component<FXViewProps, FXViewState> {
  state: FXViewState = {
    components: [],
  };

  private viewId: string;

  constructor(props: FXViewProps) {
    super(props);
    // 使用 fxViewId，如果没有则生成一个唯一 ID
    this.viewId =
      props.fxViewId ||
      `auto_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
  }

  componentDidMount() {
    logger.log("FXView", `mounted with viewId: ${this.viewId}`);
    // 注册到视图管理器
    FXManager.registerView(this.viewId, this.updateComponents);
    // 初始更新
    this.updateComponents();
  }

  componentDidUpdate(_prevProps: FXViewProps) {
    logger.log("FXView", `updated: ${this.viewId}`);
    const newViewId = this.props.fxViewId || this.viewId;

    if (newViewId !== this.viewId) {
      // ID 变化时重新注册
      logger.log("FXView", `updated changed: ${this.viewId}`);
      FXManager.unregisterView(this.viewId);
      this.viewId = newViewId;

      FXManager.registerView(this.viewId, this.updateComponents);
      this.updateComponents();
    }
  }

  componentWillUnmount() {
    logger.log("FXView", `unmounted: ${this.viewId}`);
    // 注销视图（会自动清理所有组件和回调）
    FXManager.unregisterView(this.viewId);
  }

  updateComponents = () => {
    const components = FXManager.getComponents(this.viewId);
    this.setState({ components: components });
  };

  render() {
    const { children } = this.props;
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { fxViewId, ...restProps } = this.props;
    const { components } = this.state;
    logger.log("FXView", `rendering components: ${components.length}`);
    return (
      <View {...restProps}>
        {children}
        {components.map((item) =>
          item.visible ? (
            <React.Fragment key={item.componentId}>
              {item.component}
            </React.Fragment>
          ) : null,
        )}
      </View>
    );
  }
}

export type { FXViewProps };
