// react-native-fxview/FXViewController.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FXViewController = void 0;
const FXCategoryController_1 = require("./FXCategoryController");
const FXLogger_1 = __importDefault(require("./logger/FXLogger"));
class FXViewController {
    constructor(fxViewId) {
        // 存储各分类的控制器
        this.categoryControllerMap = new Map();
        this.fxViewId = fxViewId;
    }
    //#region ========== 主要 API 方法 ==========
    /**
     * 创建并显示组件
     * @param component 组件内容
     * @param categoryId 分类 ID（可选，默认 "default"）
     * @param componentId 组件 ID（可选，自动生成）
     * @returns 组件控制器
     */
    add(component, categoryId, componentId) {
        FXLogger_1.default.info("FXViewController.add", `[${this.fxViewId}]`, {
            categoryId,
            componentId,
        });
        const categoryController = this.getCategoryControllerOrCreate(categoryId);
        return categoryController.add(component, componentId);
    }
    /**
     * 创建但不显示组件
     * @param component 组件内容
     * @param categoryId 分类 ID（可选，默认 "default"）
     * @param componentId 组件 ID（可选，自动生成）
     * @returns 组件控制器
     */
    build(component, categoryId, componentId) {
        FXLogger_1.default.info("FXViewController.build", `[${this.fxViewId}]`, {
            categoryId,
            componentId,
        });
        const categoryController = this.getCategoryControllerOrCreate(categoryId);
        return categoryController.build(component, componentId);
    }
    /**
     * 显示已存在的组件
     * @param componentId 组件 ID
     * @param categoryId 分类 ID（可选，默认 "default"）
     */
    show(componentId, categoryId) {
        FXLogger_1.default.info("FXViewController.show", `[${this.fxViewId}]`, {
            categoryId,
            componentId,
        });
        const categoryController = this.getCategoryController(categoryId);
        if (!categoryController) {
            FXLogger_1.default.warn(`CategoryController ${categoryId || "default"} not found in view ${this.fxViewId}`);
            return;
        }
        categoryController.show(componentId);
    }
    /**
     * 隐藏但不删除组件
     * @param componentId 组件 ID
     * @param categoryId 分类 ID（可选，默认 "default"）
     */
    hide(componentId, categoryId) {
        FXLogger_1.default.info("FXViewController.hide", `[${this.fxViewId}]`, {
            categoryId,
            componentId,
        });
        const categoryController = this.getCategoryController(categoryId);
        if (!categoryController) {
            FXLogger_1.default.warn(`CategoryController ${categoryId || "default"} not found in view ${this.fxViewId}`);
            return;
        }
        categoryController.hide(componentId);
    }
    /**
     * 更新组件内容
     * @param componentId 组件 ID
     * @param component 新的组件内容
     * @param categoryId 分类 ID（可选，默认 "default"）
     */
    update(componentId, component, categoryId) {
        FXLogger_1.default.info("FXViewController.update", `[${this.fxViewId}]`, {
            categoryId,
            componentId,
        });
        const categoryController = this.getCategoryController(categoryId);
        if (!categoryController) {
            FXLogger_1.default.warn(`CategoryController ${categoryId || "default"} not found in view ${this.fxViewId}`);
            return;
        }
        categoryController.update(componentId, component);
    }
    /**
     * 彻底删除组件
     * @param componentId 组件 ID
     * @param categoryId 分类 ID（可选，默认 "default"）
     */
    remove(componentId, categoryId) {
        FXLogger_1.default.info("FXViewController.remove", `[${this.fxViewId}]`, {
            categoryId,
            componentId,
        });
        const categoryController = this.getCategoryController(categoryId);
        if (!categoryController) {
            FXLogger_1.default.warn(`CategoryController ${categoryId || "default"} not found in view ${this.fxViewId}`);
            return;
        }
        categoryController.remove(componentId);
        // 如果该分类没有组件了，删除整个分类控制器
        if (categoryController.getComponentCount() === 0) {
            this.removeCategory(categoryId);
        }
    }
    /**
     * 移除最后一个组件（向后兼容）
     * @param categoryId 分类 ID（可选，默认 "default"）
     */
    removeLast(categoryId) {
        FXLogger_1.default.info("FXViewController.removeLast", `[${this.fxViewId}]`, {
            categoryId,
        });
        const categoryController = this.getCategoryController(categoryId);
        if (!categoryController) {
            FXLogger_1.default.warn(`CategoryController ${categoryId || "default"} not found in view ${this.fxViewId}`);
            return;
        }
        categoryController.removeLast();
        // 如果该分类没有组件了，删除整个分类控制器
        if (categoryController.getComponentCount() === 0) {
            this.removeCategory(categoryId);
        }
    }
    /**
     * 清空所有分类的所有组件
     */
    clearAll() {
        FXLogger_1.default.info("FXViewController.clearAll", `[${this.fxViewId}]`);
        this.categoryControllerMap.forEach((controller) => {
            controller.clearAll();
        });
        this.categoryControllerMap.clear();
        FXLogger_1.default.info("FXViewController.clearAll done", this.categoryControllerMap.size);
    }
    /**
     * 清空指定分类的所有组件
     * @param categoryId 分类 ID
     */
    clearCategory(categoryId) {
        const finalCategoryId = categoryId || "default";
        FXLogger_1.default.info("FXViewController.clearCategory", `[${this.fxViewId}]`, {
            categoryId: finalCategoryId,
        });
        const categoryController = this.getCategoryController(categoryId);
        if (!categoryController) {
            FXLogger_1.default.warn(`CategoryController ${finalCategoryId} not found in view ${this.fxViewId}`);
            return;
        }
        categoryController.clearAll();
        this.categoryControllerMap.delete(finalCategoryId);
    }
    /**
     * 注册更新回调
     * @param callback 更新回调函数
     */
    registerUpdateCallback(callback) {
        this.updateCallback = callback;
        // 将回调传递给所有已存在的分类控制器
        this.categoryControllerMap.forEach((controller) => {
            controller.registerUpdateCallback(callback);
        });
    }
    /**
     * 获取组件列表
     * @param filterOptions 过滤选项
     * @returns 组件列表
     */
    getComponents(categoryId) {
        const result = [];
        this.categoryControllerMap.forEach((categoryController, category) => {
            // 如果指定了类别，只处理指定类别
            if (categoryId && categoryId !== category) {
                return;
            }
            const categoryComponents = categoryController.getComponents();
            result.push(...categoryComponents);
        });
        return result;
    }
    /**
     * 获取总组件数量
     * @returns 组件数量
     */
    getComponentCount() {
        let count = 0;
        this.categoryControllerMap.forEach((controller) => {
            count += controller.getComponentCount();
        });
        return count;
    }
    /**
     * 获取可见组件数量
     * @returns 可见组件数量
     */
    getVisibleComponentCount() {
        let count = 0;
        this.categoryControllerMap.forEach((controller) => {
            count += controller.getVisibleComponentCount();
        });
        return count;
    }
    /**
     * 检查组件是否存在
     * @param componentId 组件 ID
     * @param categoryId 分类 ID（可选）
     * @returns 是否存在
     */
    hasComponent(componentId, categoryId) {
        var _a;
        if (categoryId) {
            const controller = this.getCategoryController(categoryId);
            return (_a = controller === null || controller === void 0 ? void 0 : controller.hasComponent(componentId)) !== null && _a !== void 0 ? _a : false;
        }
        // 如果没有指定分类，搜索所有分类
        for (const controller of this.categoryControllerMap.values()) {
            if (controller.hasComponent(componentId)) {
                return true;
            }
        }
        return false;
    }
    /**
     * 检查组件是否可见
     * @param componentId 组件 ID
     * @param categoryId 分类 ID（可选）
     * @returns 是否可见
     */
    isComponentVisible(componentId, categoryId) {
        var _a;
        if (categoryId) {
            const controller = this.getCategoryController(categoryId);
            return (_a = controller === null || controller === void 0 ? void 0 : controller.isVisible(componentId)) !== null && _a !== void 0 ? _a : false;
        }
        // 如果没有指定分类，搜索所有分类
        for (const controller of this.categoryControllerMap.values()) {
            if (controller.hasComponent(componentId)) {
                return controller.isVisible(componentId);
            }
        }
        return false;
    }
    //#endregion
    //#region ========== 私有方法 ==========
    /**
     * 获取分类控制器
     * @param categoryId 分类 ID
     * @returns 分类控制器或 null
     */
    getCategoryController(categoryId) {
        const finalCategoryId = categoryId || "default";
        return this.categoryControllerMap.get(finalCategoryId) || null;
    }
    /**
     * 获取或创建分类控制器
     * @param categoryId 分类 ID
     * @returns 分类控制器
     */
    getCategoryControllerOrCreate(categoryId) {
        const finalCategoryId = categoryId || "default";
        let categoryController = this.categoryControllerMap.get(finalCategoryId);
        if (!categoryController) {
            categoryController = new FXCategoryController_1.FXCategoryController(this.fxViewId, finalCategoryId, this.updateCallback);
            this.categoryControllerMap.set(finalCategoryId, categoryController);
        }
        return categoryController;
    }
    /**
     * 删除分类控制器
     * @param categoryId 分类 ID
     */
    removeCategory(categoryId) {
        const finalCategoryId = categoryId || "default";
        FXLogger_1.default.info("FXViewController.removeCategory", `[${this.fxViewId}]`, {
            categoryId: finalCategoryId,
        });
        this.categoryControllerMap.delete(finalCategoryId);
    }
}
exports.FXViewController = FXViewController;
