// react-native-fxview/FXViewManager.ts
import React from "react";
import {
  FXComponentController,
  FXComponentItem,
  FXLifecycleCallbacks,
} from "./types";
import { FXViewController } from "./FXViewController";
import { HeapType, PriorityOrder, PriorityQueue } from "./queue/PriorityQueue";
import logger from "./logger/FXLogger";

export class FXViewManager {
  private static instance: FXViewManager;

  // 存储视图控制器
  private viewControllerMap: Map<string, FXViewController> = new Map();
  private fxViewIdQueue: PriorityQueue<string> = new PriorityQueue(
    HeapType.MAX_HEAP,
    PriorityOrder.LIFO,
  );
  private viewLifecycleCallbacks: Array<FXLifecycleCallbacks> = [];

  static getInstance(): FXViewManager {
    if (!FXViewManager.instance) {
      FXViewManager.instance = new FXViewManager();
    }
    return FXViewManager.instance;
  }

  //#region ========== 主要 API 方法 ==========

  /**
   * 创建并显示组件
   * @param component 组件内容
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选，默认 "default"）
   * @param componentId 组件 ID（可选，自动生成）
   * @returns 组件控制器
   */
  add(
    component: React.ReactNode,
    fxViewId?: string,
    categoryId?: string,
    componentId?: string,
  ): FXComponentController {
    logger.log("FXViewManager", "add", {
      fxViewId,
      categoryId,
      componentId,
    });

    const viewController = this.getViewControllerOrCreate(fxViewId);
    return viewController.add(component, categoryId, componentId);
  }

  /**
   * 创建但不显示组件
   * @param component 组件内容
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选，默认 "default"）
   * @param componentId 组件 ID（可选，自动生成）
   * @returns 组件控制器
   */
  build(
    component: React.ReactNode,
    fxViewId?: string,
    categoryId?: string,
    componentId?: string,
  ): FXComponentController {
    logger.log("FXViewManager", "build", {
      fxViewId,
      categoryId,
      componentId,
    });

    const viewController = this.getViewControllerOrCreate(fxViewId);
    return viewController.build(component, categoryId, componentId);
  }

  /**
   * 显示已存在的组件
   * @param componentId 组件 ID
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选，默认 "default"）
   */
  show(componentId: string, fxViewId?: string, categoryId?: string): void {
    logger.log("FXViewManager", "show", {
      fxViewId,
      categoryId,
      componentId,
    });

    const viewController = this.getViewController(fxViewId);
    if (!viewController) {
      logger.warn("FXViewManager", `FXView ${fxViewId || "latest"} not found`);
      return;
    }

    viewController.show(componentId, categoryId);
  }

  /**
   * 隐藏但不删除组件
   * @param componentId 组件 ID
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选，默认 "default"）
   */
  hide(componentId: string, fxViewId?: string, categoryId?: string): void {
    logger.log("FXViewManager", "hide", {
      fxViewId,
      categoryId,
      componentId,
    });

    const viewController = this.getViewController(fxViewId);
    if (!viewController) {
      logger.warn("FXViewManager", `FXView ${fxViewId || "latest"} not found`);
      return;
    }

    viewController.hide(componentId, categoryId);
  }

  /**
   * 更新组件内容
   * @param componentId 组件 ID
   * @param component 新的组件内容
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选，默认 "default"）
   */
  update(
    componentId: string,
    component: React.ReactNode,
    fxViewId?: string,
    categoryId?: string,
  ): void {
    logger.log("FXViewManager", "update", {
      fxViewId,
      categoryId,
      componentId,
    });

    const viewController = this.getViewController(fxViewId);
    if (!viewController) {
      logger.warn("FXViewManager", `FXView ${fxViewId || "latest"} not found`);
      return;
    }

    viewController.update(componentId, component, categoryId);
  }

  /**
   * 彻底删除组件
   * @param componentId 组件 ID
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选，默认 "default"）
   */
  remove(componentId: string, fxViewId?: string, categoryId?: string): void {
    logger.log("FXViewManager", "remove", {
      fxViewId,
      categoryId,
      componentId,
    });

    const viewController = this.getViewController(fxViewId);
    if (!viewController) {
      logger.warn("FXViewManager", `FXView ${fxViewId || "latest"} not found`);
      return;
    }

    viewController.remove(componentId, categoryId);
  }

  /**
   * 移除最后一个组件（向后兼容）
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选，默认 "default"）
   */
  removeLast(fxViewId?: string, categoryId?: string): void {
    logger.log("FXViewManager", "removeLast", {
      fxViewId,
      categoryId,
    });

    const viewController = this.getViewController(fxViewId);
    if (!viewController) {
      logger.warn("FXViewManager", `FXView ${fxViewId || "latest"} not found`);
      return;
    }

    viewController.removeLast(categoryId);
  }

  /**
   * 清空指定视图的所有组件
   * @param fxViewId 视图 ID（可选，清空所有视图）
   */
  clearAll(fxViewId?: string): void {
    logger.log("FXViewManager", "clearAll", { fxViewId });

    if (fxViewId) {
      // 清空指定视图
      const viewController = this.getViewController(fxViewId);
      if (viewController) {
        viewController.clearAll();
      }
    } else {
      // 清空所有视图
      this.viewControllerMap.forEach((controller) => {
        controller.clearAll();
      });
    }
  }

  /**
   * 清空指定分类的所有组件
   * @param categoryId 分类 ID
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   */
  clearCategory(categoryId: string, fxViewId?: string): void {
    logger.log("FXViewManager", "clearCategory", { fxViewId, categoryId });

    const viewController = this.getViewController(fxViewId);
    if (!viewController) {
      logger.warn("FXViewManager", `FXView ${fxViewId || "latest"} not found`);
      return;
    }

    viewController.clearCategory(categoryId);
  }

  //#endregion

  //#region ========== 视图管理方法 ==========

  /**
   * 注册视图（FXView 挂载时调用）
   * @param fxViewId 视图 ID
   * @param updateCallback 更新回调函数
   * @returns 视图控制器
   */
  registerView(fxViewId: string, updateCallback: () => void): FXViewController {
    logger.log("FXViewManager", "registerView", fxViewId);

    let viewController = this.viewControllerMap.get(fxViewId);

    if (!viewController) {
      viewController = new FXViewController(fxViewId);
      this.viewControllerMap.set(fxViewId, viewController);
    }

    // 注册更新回调
    viewController.registerUpdateCallback(updateCallback);
    this.fxViewIdQueue.remove(fxViewId);
    this.fxViewIdQueue.enqueue(fxViewId);
    this.viewLifecycleCallbacks.forEach((callback) => {
      callback.didMount?.(fxViewId);
    });
    return viewController;
  }

  /**
   * 注销视图（FXView 卸载时调用）
   * @param fxViewId 视图 ID
   * @returns 是否成功注销
   */
  unregisterView(fxViewId: string): boolean {
    logger.log("FXViewManager", "unregisterView", fxViewId);

    const viewController = this.getViewController(fxViewId);
    if (!viewController) {
      logger.warn(
        "FXViewManager",
        `FXView ${fxViewId} not found, cannot unregister`,
      );
      return false;
    }

    // 清空回调
    viewController.registerUpdateCallback(undefined);

    // 清空所有组件
    viewController.clearAll();

    // 从队列中移除
    this.fxViewIdQueue.remove(fxViewId);

    // 从 Map 中删除
    this.viewControllerMap.delete(fxViewId);

    this.viewLifecycleCallbacks.forEach((callback) => {
      callback.willUnmount?.(fxViewId);
    });

    return true;
  }

  /**
   * 注册生命周期回调
   * @param callbacks 生命周期回调
   */
  registerLifecycleCallbacks(callbacks: FXLifecycleCallbacks): void {
    logger.log("FXViewManager", "registerLifecycleCallbacks");
    this.viewLifecycleCallbacks.push(callbacks);
  }

  /**
   * 注销生命周期回调
   * @param callbacks 生命周期回调
   */
  unregisterLifecycleCallbacks(callbacks: FXLifecycleCallbacks): void {
    logger.log("FXViewManager", "unregisterLifecycleCallbacks");
    this.viewLifecycleCallbacks = this.viewLifecycleCallbacks.filter(
      (callback) => callback !== callbacks,
    );
  }

  /**
   * 获取最近使用的视图 ID（栈顶）
   * @returns 最近的视图 ID 或 null
   */
  getLatestFXViewId(): string | null {
    return this.fxViewIdQueue.peek() || null;
  }

  /**
   * 获取组件列表
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选）
   * @returns 组件列表
   */
  getComponents(fxViewId?: string, categoryId?: string): FXComponentItem[] {
    logger.log("FXViewManager", "getComponents", { fxViewId, categoryId });

    const finalFXViewId = fxViewId || this.getLatestFXViewId();
    if (!finalFXViewId) {
      logger.warn("FXViewManager", "No FXView available");
      return [];
    }

    const viewController = this.getViewController(finalFXViewId);
    if (!viewController) {
      logger.warn(
        "FXViewManager",
        `FXViewController ${finalFXViewId} not found`,
      );
      return [];
    }

    return viewController.getComponents(categoryId);
  }

  /**
   * 获取组件数量
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @returns 组件数量
   */
  getComponentCount(fxViewId?: string): number {
    const viewController = this.getViewController(fxViewId);
    return viewController?.getComponentCount() ?? 0;
  }

  /**
   * 获取可见组件数量
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @returns 可见组件数量
   */
  getVisibleComponentCount(fxViewId?: string): number {
    const viewController = this.getViewController(fxViewId);
    return viewController?.getVisibleComponentCount() ?? 0;
  }

  /**
   * 检查组件是否存在
   * @param componentId 组件 ID
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选）
   * @returns 是否存在
   */
  hasComponent(
    componentId: string,
    fxViewId?: string,
    categoryId?: string,
  ): boolean {
    const viewController = this.getViewController(fxViewId);
    return viewController?.hasComponent(componentId, categoryId) ?? false;
  }

  /**
   * 检查组件是否可见
   * @param componentId 组件 ID
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @param categoryId 分类 ID（可选）
   * @returns 是否可见
   */
  isComponentVisible(
    componentId: string,
    fxViewId?: string,
    categoryId?: string,
  ): boolean {
    const viewController = this.getViewController(fxViewId);
    return viewController?.isComponentVisible(componentId, categoryId) ?? false;
  }

  //#endregion

  //#region ========== 私有方法 ==========

  /**
   * 获取视图控制器
   * @param fxViewId 视图 ID（可选，使用最近的视图）
   * @returns 视图控制器或 null
   */
  private getViewController(fxViewId?: string): FXViewController | null {
    const finalFXViewId = fxViewId || this.getLatestFXViewId();
    if (!finalFXViewId) {
      logger.warn("FXViewManager", "getViewController: no fxViewId available");
      return null;
    }
    return this.viewControllerMap.get(finalFXViewId) || null;
  }

  /**
   * 获取或创建视图控制器
   * @param fxViewId 视图 ID（可选，使用最近的视图或创建新的）
   * @returns 视图控制器
   */
  private getViewControllerOrCreate(fxViewId?: string): FXViewController {
    const finalFXViewId =
      fxViewId || this.getLatestFXViewId() || this.autoFXViewId();

    let viewController = this.viewControllerMap.get(finalFXViewId);

    if (!viewController) {
      logger.warn(
        "FXViewManager",
        `FXViewController ${finalFXViewId} not registered, creating a temporary one`,
      );
      viewController = new FXViewController(finalFXViewId);
      this.viewControllerMap.set(finalFXViewId, viewController);
    }

    return viewController;
  }

  /**
   * 自动生成 fxViewId
   * @returns 生成的 fxViewId
   */
  private autoFXViewId(): string {
    return `fxView_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
  }

  //#endregion
}

const FXManager = FXViewManager.getInstance();
export default FXManager;
