// react-native-fxview/FXView.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const react_native_1 = require("react-native");
const FXViewManager_1 = __importDefault(require("./FXViewManager"));
const FXLogger_1 = __importDefault(require("./logger/FXLogger"));
class FXView extends react_1.Component {
    constructor(props) {
        super(props);
        this.state = {
            components: [],
        };
        this.updateComponents = () => {
            const components = FXViewManager_1.default.getComponents(this.viewId);
            this.setState({ components: components });
        };
        // 使用 fxViewId，如果没有则生成一个唯一 ID
        this.viewId =
            props.fxViewId ||
                `auto_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    }
    componentDidMount() {
        FXLogger_1.default.log("FXView", `mounted with viewId: ${this.viewId}`);
        // 注册到视图管理器
        FXViewManager_1.default.registerView(this.viewId, this.updateComponents);
        // 初始更新
        this.updateComponents();
    }
    componentDidUpdate(_prevProps) {
        FXLogger_1.default.log("FXView", `updated: ${this.viewId}`);
        const newViewId = this.props.fxViewId || this.viewId;
        if (newViewId !== this.viewId) {
            // ID 变化时重新注册
            FXLogger_1.default.log("FXView", `updated changed: ${this.viewId}`);
            FXViewManager_1.default.unregisterView(this.viewId);
            this.viewId = newViewId;
            FXViewManager_1.default.registerView(this.viewId, this.updateComponents);
            this.updateComponents();
        }
    }
    componentWillUnmount() {
        FXLogger_1.default.log("FXView", `unmounted: ${this.viewId}`);
        // 注销视图（会自动清理所有组件和回调）
        FXViewManager_1.default.unregisterView(this.viewId);
    }
    render() {
        const { children } = this.props;
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const _a = this.props, { fxViewId } = _a, restProps = __rest(_a, ["fxViewId"]);
        const { components } = this.state;
        FXLogger_1.default.log("FXView", `rendering components: ${components.length}`);
        return (<react_native_1.View {...restProps}>
        {children}
        {components.map((item) => item.visible ? (<react_1.default.Fragment key={item.componentId}>
              {item.component}
            </react_1.default.Fragment>) : null)}
      </react_native_1.View>);
    }
}
exports.default = FXView;
